test/extended/router/externalCertificate.go

```diff
@@ -0,0 +1,682 @@
+						o.Expect(err).NotTo(o.HaveOccurred())
+					})
+					g.It("route update is rejcted", func() {
+						g.By("Updating the route to use new external certificate")
+						routePatch := fmt.Sprintf(`{"spec":{"tls":{"externalCertificate":{"name":"%s"}}}}`, newSecretName)
+						_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Patch(
+							context.Background(), routeToTest.Name, types.MergePatchType, []byte(routePatch), metav1.PatchOptions{},
+						)
+						o.Expect(err).To(o.HaveOccurred())
+						o.Expect(err.Error()).To(o.And(
+							o.ContainSubstring("Forbidden: router serviceaccount does not have permission to get this secret"),
+							o.ContainSubstring("Forbidden: router serviceaccount does not have permission to watch this secret"),
+							o.ContainSubstring("Forbidden: router serviceaccount does not have permission to list this secret")),
+						)
+					})
+				})
+				g.Context("to use new external certificate, but secret is not of type kubernetes.io/tls", func() {
+					g.BeforeEach(func() {
+						g.By("Creating a secret with the WRONG type (Opaque)")
+						secret, _, err := createSecret(oc.Namespace(), newSecretName, corev1.SecretTypeOpaque, hosts...) // Incorrect type
+						o.Expect(err).NotTo(o.HaveOccurred())
+						_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+						o.Expect(err).NotTo(o.HaveOccurred())
+					})
+					g.It("route update is rejcted", func() {
+						g.By("Updating the route to use new external certificate")
+						routePatch := fmt.Sprintf(`{"spec":{"tls":{"externalCertificate":{"name":"%s"}}}}`, newSecretName)
+						_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Patch(
+							context.Background(), routeToTest.Name, types.MergePatchType, []byte(routePatch), metav1.PatchOptions{},
+						)
+						o.Expect(err).To(o.HaveOccurred())
+						o.Expect(err.Error()).To(o.ContainSubstring(fmt.Sprintf(`Invalid value: "%s": secret of type "kubernetes.io/tls" required`, newSecretName)))
+					})
+
+				})
+				g.Context("to use new external certificate, but secret does not exist", func() {
+					// do not create new secret
+					g.It("route update is rejcted", func() {
+						g.By("Updating the route to use new external certificate")
+						routePatch := fmt.Sprintf(`{"spec":{"tls":{"externalCertificate":{"name":"%s"}}}}`, newSecretName)
+						_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Patch(
+							context.Background(), routeToTest.Name, types.MergePatchType, []byte(routePatch), metav1.PatchOptions{},
+						)
+						o.Expect(err).To(o.HaveOccurred())
+						o.Expect(err.Error()).To(o.ContainSubstring(fmt.Sprintf(`Not found: "secrets \"%s\" not found"`, newSecretName)))
+					})
+				})
+				g.Context("to use same external certificate", func() {
+					g.BeforeEach(func() {
+						g.By("Adding some label to trigger route update")
```

Comment:
QQ(implementation): Certificate Renewal related

haven't been following up to work my way up all the discussions you've had in this regard,
I know from past discussions that an update event is necessary to trigger any update - but does this the Certificate secret contents being updated when renewal changes the contents (I think the singleitemlistwatch should take care)?

Reply:
Right, Certificate Renewal will update the contents of the secret which will send an update event to the secret monitor (singleitemlistwatch). 

The following test is added to test the scenario

```
"[sig-network][OCPFeatureGate:RouteExternalCertificate][Feature:Router][apigroup:route.openshift.io] with valid setup the router should support external certificate and the secret is updated then also routes are reachable": " [Suite:openshift/conformance/parallel]",
```

Reply:
The above test (against which you have commented) is added to test scenarios where route will be updated but `ExternalCertificate` field will remain unchanged -- in that case also, we want to do the validation of the feature, so that dropping RBAC permissions will get checked. 

The following test is added for this:
```
"[sig-network][OCPFeatureGate:RouteExternalCertificate][Feature:Router][apigroup:route.openshift.io] with valid setup the router should support external certificate and the route is updated to use same external certificate, but RBAC permissions are dropped route update is rejected": " [Suite:openshift/conformance/parallel]",

```